nvflare/app_opt/tensor_stream/receiver.py

```diff
@@ -0,0 +1,199 @@
+        super().__init__()
+        self.engine = engine
+        self.ctx_prop_key = ctx_prop_key
+        self.format = format
+        self.channel = channel
+        # key: task_id, value: tensors received from the peer
+        self.tensors: dict[str, TensorsMap] = {}
+        self.tensor_events = {}  # Maps task_id to Event objects
+        self.lock = threading.Lock()
+        self.logger = get_obj_logger(self)
+        self._register()
+
+    def _register(self):
+        """Register the consumer factory with the engine."""
+        topic = get_topic_for_ctx_prop_key(self.ctx_prop_key)
+        self.engine.register_stream_processing(
+            channel=self.channel,
+            topic=topic,
+            factory=TensorConsumerFactory(),
+            stream_done_cb=self._save_tensors_cb,
+        )
+        self.logger.debug(
+            f"Registered tensor receiver for context property '{self.ctx_prop_key}' "
+            f"on '{self.channel}:{topic}' with format '{self.format}'.",
+        )
+
+    def _save_tensors_cb(self, success: bool, fl_ctx: FLContext):
+        """Save tensors received from stream. Called when the stream is done.
+
+        Args:
+            success (bool): no error happen on the stream consumer
+            fl_ctx (FLContext): the FLContext for the current operation
+        """
+        peer_name = fl_ctx.get_peer_context().get_identity_name()
+        task_id = fl_ctx.get_custom_prop(TensorCustomKeys.TASK_ID)
+        if not task_id:
+            raise ValueError(f"No task_id found from peer {peer_name}.")
+
+        with self.lock:
+            if task_id not in self.tensor_events:
+                self.tensor_events[task_id] = threading.Event()
+
+        if not success:
+            with self.lock:
+                self.tensor_events[task_id].set()  # Wake waiting threads
+            raise ValueError(f"Failed to receive tensors from peer '{peer_name}' and task '{task_id}'.")
+
+        tensors = fl_ctx.get_custom_prop(TensorCustomKeys.SAFE_TENSORS_PROP_KEY)
+        if not tensors:
+            raise ValueError(f"No tensors found from peer '{peer_name}' and task '{task_id}'.")
+
+        # Clean up custom properties to reduce memory usage
+        fl_ctx.set_custom_prop(TensorCustomKeys.SAFE_TENSORS_PROP_KEY, None)
+        fl_ctx.set_custom_prop(TensorCustomKeys.TASK_ID, None)
+
+        self.on_tensor_received(task_id, tensors)
+        del tensors
+
+    def set_ctx_with_tensors(self, fl_ctx: FLContext):
+        """Update the context with the received tensors.
+
+        Args:
+            fl_ctx (FLContext): The FLContext for the current operation.
+        """
+        peer_name = fl_ctx.get_peer_context().get_identity_name()
+        task_id = fl_ctx.get_prop(FLContextKey.TASK_ID, None)
+        if not task_id:
+            raise ValueError("No task_id found in FLContext.")
+
+        # get and remove the tensors from the local store
+        tensors = self.tensors.pop(task_id, None)
```

Comment:
could there be two threads doing this at the same time?
do we need to put this inside a lock?